### PR TITLE
Fix TinyMCE ext-buttons. Respect ext-buttons configuration per an editor instance

### PR DESCRIPTION
**Pull request by Fedik**
Pull Request for Issue #14417.

### Summary of Changes
This pull restore possibility to hide the ext-buttons per TinyMCE instance.

### Testing Instructions
In the article form XML `administrator/components/com_content/models/forms/article.xml` add one more editor element (somewhere after line `<fieldset name="basic" label="COM_CONTENT_ATTRIBS_FIELDSET_LABEL">`)
```xml
<field name="text2" type="editor" hide="menu,module,contact"
 label="Text2" description="" filter="JComponentHelper::filterText" buttons="true" />
```

### Expected result
Now open the Article editing, you should got:
- main editor contain all available ex-buttons
- the additional editor do not have menu,module,contact buttons.

### Actual result
Always displayed all buttons.

### Documentation Changes Required
nope

### CODE REVIEW

File: media/editors/tinymce/js/tinymce.js [hunk 1]
```diff
@@ -49,15 +49,28 @@
...
+				options = Joomla.extend({}, options);
+			}
 
 			if (element) {
 				// We already have the Target, so reset the selector and assign given element as target
 				options.selector = null;
 				options.target   = element;
 			}
 
+			// @TODO: the ext-buttons should be as TinyMCE plugins, not the callback hack
```
Comment (dgt41) [inline media/editors/tinymce/js/tinymce.js]: That was rejected by the PLT...

**Review by dgrammatiko — APPROVED** [review]

**Comment by franz-wohlkoenig** [comment]
I have tested this item :red_circle: unsuccessfully on 7efc54b61aac3955eb8e955004653cad666ce8b9<br /><br />In none Tab a `text2`-Editor is shown.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/14520">issues.joomla.org/tracker/joomla-cms/14520</a>.</sub>

**Comment by franz-wohlkoenig** [comment]
I have tested this item :white_check_mark: successfully on 7efc54b61aac3955eb8e955004653cad666ce8b9<br /><br />On Tab `Options` Editor `Text2` is shown (first changes in xml-File got not saved).<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/14520">issues.joomla.org/tracker/joomla-cms/14520</a>.</sub>

**Comment by Bakual** [comment]
@Fedik Can you fix the conflicts in your PR please?

**Comment by Bakual** [comment]
I have tested this item :white_check_mark: successfully on 7efc54b61aac3955eb8e955004653cad666ce8b9<br /><br />This fixes the editor buttons for when they are set to hidden in the form xml. It also fixes the issue #14935 for that case.
However both issues remain for the case where the editor is added using a custom editor field.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/14520">issues.joomla.org/tracker/joomla-cms/14520</a>.</sub>

**Comment by Fedik** [comment]
@Bakual conflict fixed

**Comment by Bakual** [comment]
Looks like it works now also for the custom fields. Either your conflict solving or me changing some settings in my testing site fixed it.
All good now as far as I see.

**Comment by Bakual** [comment]
I have tested this item :white_check_mark: successfully on 09b6ad3a7fa6b703b294aabbde1023710d65bd16<br /><br />Tested after conflict solving again.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/14520">issues.joomla.org/tracker/joomla-cms/14520</a>.</sub>

**Comment by laoneo** [comment]
I will test that pr tomorrow morning as well.

**Comment by Fedik** [comment]
please remove RTC, 
later I try to fis #14935 also, it is somehow related

**Comment by franz-wohlkoenig** [comment]
Remove RTC as wanted by @Fedik <hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/14520">issues.joomla.org/tracker/joomla-cms/14520</a>.</sub>

**Comment by PhilETaylor** [comment]
There are also merge conflicts now

**Comment by Bakual** [comment]
> There are also merge conflicts now

That's not true:

> This branch has no conflicts with the base branch

**Comment by PhilETaylor** [comment]
Well the patch file doesnt apply, it rejects... https://patch-diff.githubusercontent.com/raw/joomla/joomla-cms/pull/14520.patch

**Comment by Fedik** [comment]
> There are also merge conflicts now

it is a last problem here 😄

**Comment by Fedik** [comment]
Please test now again.
+
Test these issues which also should be fixed: #14935  #14990

**Comment by Bakual** [comment]
I have tested this item :white_check_mark: successfully on 96603d295c396c173cb22825576af8d6ca4bdf0c<br /><br />Tested both with my extension and a custom editor field.
* Show Button setting is now respected both in form.xml and custom field parameter
* Hide Button setting is now respected in both places as well and it also works when the parameter is empty
* Buttons work as expected.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/14520">issues.joomla.org/tracker/joomla-cms/14520</a>.</sub>

**Comment by PhilETaylor** [comment]
This fixes #14935 #14990

BUT - incoming, another bug for you to fix - will log it now. This PR can be improved to fix that next bug.

**Comment by dgt41** [comment]
I have tested this item :white_check_mark: successfully on 96603d295c396c173cb22825576af8d6ca4bdf0c<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/14520">issues.joomla.org/tracker/joomla-cms/14520</a>.</sub>

**Comment by dgt41** [comment]
Thank you @Fedik 

RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/14520">issues.joomla.org/tracker/joomla-cms/14520</a>.</sub>

**Comment by PhilETaylor** [comment]
Hold off on that RTC please :)

**Comment by PhilETaylor** [comment]
Please make this PR fix this multiple editor bug too: https://github.com/joomla/joomla-cms/issues/15014

**Comment by zero-24** [comment]
@PhilETaylor i think we can fix the other bug in a new PR. I'm merging this for now and we can fix the other bug with a new PR.

Thanks! 👍

**Comment by dgt41** [comment]
@PhilETaylor the code is not related to tinyMCE: https://github.com/joomla/joomla-cms/blob/staging/media/com_content/js/admin-article-readmore.js

**Comment by PhilETaylor** [comment]
All I know is it doesn't work, Im in idiot mode this week, Im not programming this week for Joomla at all, just acting like an idiot and reporting bugs - tomorrow I leave the office for 3 weeks Easter holiday to Singapore with my 10 year old daughter, so i'm winding down. I'll report them, you fix them :)

**Comment by dgt41** [comment]
``js 
content = (new Function('return ' + options.editor))();
```
replace it with:
```js
if (window.parent.Joomla && window.parent.Joomla.editors && window.parent.Joomla.editors.instances && window.parent.Joomla.editors.instances.hasOwnProperty(editor)) {
			content = window.parent.Joomla.editors.instances[editor].getContent()
		} else {
			content = (new Function('return ' + options.editor))();
		}
```

Simple fix

**Comment by laoneo** [comment]
Good work!

**Comment by laoneo** [comment]
Would that mean we can have now different editors on the same page? Then we can think about a setting in the editor custom field to define which editor should be loaded.

**Comment by Bakual** [comment]
> Would that mean we can have now different editors on the same page? Then we can think about a setting in the editor custom field to define which editor should be loaded.

As I understood @dgt41, different editors on the same page work if all editor-xtd plugins are rewritten to the new format. It doesn't work for those using the old format (which will be a lot 3rd party ones). So for now I wouldn't add the option to specify the editor type. With 4.0 we can add it.

**Comment by dgt41** [comment]
> Would that mean we can have now different editors on the same page?

Not so soon :( 
The problem is that all the existing editors (apart the core ones) are still using the old `jInsertEditorText()` which by design is **not** supporting multiple editors (different type).
On top of that as @Bakual wrote above **all** the XTDs are using as well the old method `jInsertEditorText()`.

The only thing we could do is prepare all the plugins to use the new API if the editor supports it e.g.
```js
if (window.parent.Joomla && window.parent.Joomla.editors && window.parent.Joomla.editors.instances && window.parent.Joomla.editors.instances.hasOwnProperty(editor)) {
	content = window.parent.Joomla.editors.instances[editor].getContent()
} else {
	content = (new Function('return ' + options.editor))();
}
```
Also there is another instance that needs to follow the API in /administarator/components/com_fields/view/fields/tmpl/modal.php
instead of 
```js
fieldIns = function(id) {
	window.parent.jInsertEditorText("{field " + id + "}", "' . $editor . '");
	window.parent.jModalClose();
};
fieldgroupIns = function(id) {
	window.parent.jInsertEditorText("{fieldgroup " + id + "}", "' . $editor . '");
	window.parent.jModalClose();
};
```
we should use 
```js
fieldIns = function(id) {
if (window.parent.Joomla && window.parent.Joomla.editors && window.parent.Joomla.editors.instances && window.parent.Joomla.editors.instances.hasOwnProperty(editor)) {
	window.parent.Joomla.editors.instances[editor].replaceSelection("{field " + id + "}")
} else {
	window.parent.jInsertEditorText("{field " + id + "}", "' . $editor . '");
}
	window.parent.jModalClose();
};
fieldgroupIns = function(id) {
if (window.parent.Joomla && window.parent.Joomla.editors && window.parent.Joomla.editors.instances && window.parent.Joomla.editors.instances.hasOwnProperty(editor)) {
        window.parent.Joomla.editors.instances[editor].replaceSelection("{fieldgroup " + id + "}")
} else {
	window.parent.jInsertEditorText("{fieldgroup " + id + "}", "' . $editor . '");
}
	window.parent.jModalClose();
};
```

**Comment by dgt41** [comment]
@PhilETaylor if we apply these changes to read more and fields modal view then **all** core components are supporting multiple (different types) editors per page.
But all 3rd PD editors and XTDs will **not** if they don't update their code to use the API

**Comment by PhilETaylor** [comment]
which is exactly how it should be :)

**Comment by dgt41** [comment]
ok then, someone needs to copy paste those code snippets and create the PRs

**Pull request merged**